shared/js/moz_intl.js

```diff
@@ -71,24 +71,32 @@ global.mozIntl = {
   DateTimeFormat: function(locales, options) {
     const resolvedOptions = Object.assign({}, options);
 
-    if (options.dayperiod) {
-      if (!('hour' in options)) {
+    if (resolvedOptions.dayperiod) {
+      if (resolvedOptions.hour === undefined) {
         resolvedOptions.hour = 'numeric';
       }
     }
 
-    if (resolvedOptions.hour &&
-        !options.hasOwnProperty('hour12')) {
+    if (resolvedOptions.hour !== undefined &&
+        resolvedOptions.hour12 === undefined) {
       resolvedOptions.hour12 = navigator.mozHour12;
     }
 
+    if (resolvedOptions.dayperiod === undefined &&
+        resolvedOptions.hour12 === true) {
+      resolvedOptions.dayperiod = true;
+    }
+
     var intlFormat = new Intl.DateTimeFormat(locales, resolvedOptions);
 
+    resolvedOptions.locale = intlFormat.resolvedOptions().locale;
+    resolvedOptions.hour12 = intlFormat.resolvedOptions().hour12;
+
     // This is needed for a workaround for bug 1208808
     // Remove when that bug is fixed
     var hourFormatter;
-    if (options.dayperiod === true || options.dayperiod === false &&
-        intlFormat.resolvedOptions().hour12 === true) {
+    if (resolvedOptions.dayperiod !== undefined &&
```

Comment:
I don't think so. My logic is (and it's preliminary for now, since I haven't had time to write a spec proposal yet), that:

1) If dayperiod is specified, hour has to be
2) if hour is specified and hour12 is not, we need to resolve it from mozHour12
3) if dayperiod is not specified and hour12 is true, dayperiod is true

and then for one scenario:

4) if dayperiod is specified and hour12 is true, we need additional formatter to format dayperiod potentially.

I don't see how we can combine them, and even if it's possible, I'm worried it would make it harder to understand. Since it's a "work in progress" state where we're going to start specing it soon, I like that it's fairly explicit here.
My brain capabilities are limited *^^*